File: warehouse_app/warehouse_app/report/serah_terima_gudang/serah_terima_gudang.js
// Copyright (c) 2026, Muhammad Yusuf Tri Daryanto
// License: MIT

// W4 Fase C — client script papan "Serah Terima Gudang": render + route saja.
// R1/R5: tombol tidak menulis apa pun — hanya membuka mapper native
// make_stock_entry (whitelisted, material_request.py:935-936) lewat
// frappe.model.open_mapped_doc (create_new.js:325-360 → method POST
// frappe.model.mapper.make_mapped_doc). Validasi qty/batch/over-qty tetap
// native saat SE save/submit (material_request.py:424-429). Nol endpoint
// baru, nol doc_events, nol custom field (AGENTS.md).
//
// Kolom "Aksi" didefinisikan server-side di get_columns() serah_terima_gudang.py
// (display-only, tanpa nilai). Injeksi kolom murni dari JS rapuh: hook
// get_datatable_options hanya dipakai pada branch `new window.DataTable`
// (query_report.js:1130-1136), sedangkan refresh berikutnya memakai
// datatable.refresh(data, columns) dengan kolom server saja (query_report.js:
// 1122-1126, 1076) — kolom ekstern JS akan hilang saat filter diganti.

frappe.query_reports["Serah Terima Gudang"] = {
	// Chip status = indikator baku Frappe
	// (frappe/public/scss/common/indicator.scss: .indicator-pill.<color>,
	// palet: green cyan blue orange yellow gray red dst.).
	chip_colors: {
		"Belum Dikirim": "orange",
		Sebagian: "blue",
		Terkirim: "green",
	},

	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (!data) return value; // baris total / data proxy

		if (column.fieldname === "status_papan") {
			const color = this.chip_colors[value] || "gray";
			return `<span class="indicator-pill ${color}">${frappe.utils.escape_html(
				value
			)}</span>`;
		}

		// Kolom Aksi: tombol hanya untuk bucket Belum Dikirim / Sebagian.
		// Baris Terkirim tanpa aksi — penutupan sisa MR lewat Stop MR native
		// (SOP R9), bukan kode.
		if (column.fieldname === "aksi") {
			if (data.status_papan === "Terkirim") return "";
			const mr = frappe.utils.escape_html(cstr(data.material_request));
			return `<button type="button" class="btn btn-xs btn-default w4-buat-se" data-mr="${mr}">${__(
				"Buat Stock Entry"
			)}</button>`;
		}

		return value;
	},

	onload(report) {
		// Klik didelegasikan ke $report: elemen wrapper dibuat sekali di
		// setup_report_wrapper (query_report.js init:43) dan bertahan antar
		// refresh, sementara onload dipanggil ulang tiap refresh_report
		// (query_report.js:410) → namespace + .off mencegah dobel-bind.
		report.$report
			.off("click.w4_buat_se")
			.on("click.w4_buat_se", ".w4-buat-se", function () {
				frappe.model.open_mapped_doc({
					method:
						"erpnext.stock.doctype.material_request.material_request.make_stock_entry",
					source_name: this.getAttribute("data-mr"),
				});
			});
	},
};

File: warehouse_app/warehouse_app/report/serah_terima_gudang/serah_terima_gudang.py
# Copyright (c) 2026, Muhammad Yusuf Tri Daryanto
# License: MIT

# Papan serah terima gudang (W3, Fase B).
# Catatan ref_doctype="Stock Entry": gerbang perm native query_report.run
# (frappe/desk/query_report.py:40-45) menuntut ptype "report" pada ref_doctype;
# Custom DocPerm Material Request (pemilik production_app, R7) men-set report=0
# utk semua role gudang, sedangkan Custom DocPerm Stock Entry ber-report=1 utk
# Stock User. Data report tetap Material Request; nol perubahan DocPerm.
# Grain: Material Request Item (MR + item docstatus 1, Material Transfer).
# Murni derivasi saat baca dari dokumen native: bucket status dari
# `per_ordered` native header MR; tanpa doc_events, tanpa tulis apa pun (R3/R10).
# Field custom production_app (`custom_work_order`, `custom_box_1/2`) dibaca
# display-only dan TIDAK PERNAH jadi filter/status (R4). Qty-only, tanpa valuasi.

import frappe
from frappe.utils import flt


def execute(filters=None):
	filters = frappe._dict(filters or {})

	conditions = [
		"mri.docstatus = 1",
		"mr.docstatus = 1",
		"mr.material_request_type = 'Material Transfer'",
	]
	params = {}
	if filters.get("gudang_tujuan"):
		# Filter arah gudang; nama gudang tidak di-hardcode di logika (R3).
		# Kolom native MR Item utk gudang tujuan transfer = `warehouse`
		# (`t_warehouse` hanya ada di Stock Entry Detail; mapper native
		# make_stock_entry memetakan obj.warehouse -> SE.t_warehouse).
		params["warehouse"] = resolve_warehouse(filters.gudang_tujuan)
		conditions.append("mri.warehouse = %(warehouse)s")

	data = frappe.db.sql(
		f"""
		select
			mri.parent as material_request,
			mr.transaction_date as transaction_date,
			mri.custom_work_order as work_order,
			mri.item_code as item_code,
			mri.item_name as item_name,
			mri.stock_qty as qty_diminta,
			mri.ordered_qty as qty_dikirim,
			(mri.stock_qty - mri.ordered_qty) as qty_sisa,
			mri.stock_uom as stock_uom,
			mri.warehouse as gudang_tujuan,
			mr.custom_box_1 as box_1,
			mr.custom_box_2 as box_2,
			mr.per_ordered as per_ordered,
			mr.status as status_mr
		from `tabMaterial Request Item` mri
		inner join `tabMaterial Request` mr on mr.name = mri.parent
		where {" and ".join(conditions)}
		order by mr.transaction_date desc, mr.name, mri.idx
		""",
		params,
		as_dict=1,
	)

	for row in data:
		row.status_papan = get_bucket(row.per_ordered)

	return get_columns(), data


def get_columns():
	return [
		{"label": "Material Request", "fieldname": "material_request", "fieldtype": "Link", "options": "Material Request", "width": 160},
		{"label": "Tanggal", "fieldname": "transaction_date", "fieldtype": "Date", "width": 95},
		{"label": "Work Order", "fieldname": "work_order", "fieldtype": "Link", "options": "Work Order", "width": 140},
		{"label": "Item Code", "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 160},
		{"label": "Item Name", "fieldname": "item_name", "fieldtype": "Data", "width": 180},
		{"label": "Qty Diminta", "fieldname": "qty_diminta", "fieldtype": "Float", "width": 110},
		{"label": "Sudah Dikirim", "fieldname": "qty_dikirim", "fieldtype": "Float", "width": 110},
		{"label": "Sisa", "fieldname": "qty_sisa", "fieldtype": "Float", "width": 90},
		{"label": "UOM", "fieldname": "stock_uom", "fieldtype": "Link", "options": "UOM", "width": 70},
		{"label": "Gudang Tujuan", "fieldname": "gudang_tujuan", "fieldtype": "Link", "options": "Warehouse", "width": 170},
		{"label": "Box 1", "fieldname": "box_1", "fieldtype": "Float", "width": 90},
		{"label": "Box 2", "fieldname": "box_2", "fieldtype": "Float", "width": 90},
		{"label": "Status Papan", "fieldname": "status_papan", "fieldtype": "Data", "width": 120},
		{"label": "Status MR", "fieldname": "status_mr", "fieldtype": "Data", "width": 110},
	]


def get_bucket(per_ordered):
	"""0 -> "Belum Dikirim"; 0<per_ordered<100 -> "Sebagian"; 100 -> "Terkirim".

	`per_ordered` native header MR = qty-based percent
	(status_updater.py:606-630: sum(min(ordered_qty, stock_qty)) / sum(stock_qty) * 100),
	jadi MR 10 pcs dengan SE parsial 4 => 40.0 => "Sebagian".
	"""
	per_ordered = flt(per_ordered)
	if per_ordered <= 0:
		return "Belum Dikirim"
	if per_ordered >= 100:
		return "Terkirim"
	return "Sebagian"


def resolve_warehouse(name):
	"""Resolve nama filter ke warehouse eksak; izinkan akhiran abbr situs.

	Default filter "Gudang Barang Jadi" tetap valid di situs ber-abbr (mis.
	"Gudang Barang Jadi - ROPI"): exact match dulu, kalau tidak ada, cari
	satu warehouse non-group berawalan `<nama> -`. Tanpa nama gudang yang
	di-hardcode (R3).
	"""
	if frappe.db.exists("Warehouse", name):
		return name
	match = frappe.db.get_value(
		"Warehouse",
		{"name": ("like", f"{name} -%"), "is_group": 0},
		"name",
		order_by="name",
	)
	if not match:
		frappe.throw(frappe._("Warehouse {0} tidak ditemukan").format(frappe.bold(name)))
	return match
